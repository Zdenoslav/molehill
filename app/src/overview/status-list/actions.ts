import {createActions} from 'app/redux/create-actions';

const domain = 'STATUS';

export const {
  action: deleteStatus,
} = createActions(
  `${domain}/DELETE_STATUS`,
  (id: string) => ({id}),
);

export const {
  action: deleteComment,
} = createActions(
  `${domain}/DELETE_COMMENT`,
  (id: string, statusId: string) => ({id, statusId}),
);
