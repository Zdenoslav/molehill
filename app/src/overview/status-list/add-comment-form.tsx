import React from 'react';
import { Form, InjectedFormProps, reduxForm } from 'redux-form';
import { compose } from 'redux';

import { IFormProps } from 'app/components/form-elements/typings';

import FormField from 'app/components/form-elements/form-field';
import Button from 'app/components/button';
import FormTextArea from 'app/components/form-elements/form-text-area';

export interface IAddCommentFormProps {
  body: string;
}

type Props = Partial<InjectedFormProps<IAddCommentFormProps>> & IFormProps;

class AddCommentForm extends React.PureComponent<Props> {
  public render() {
    const {
      handleSubmit,
      loading,
    } = this.props;

    return (
      <Form onSubmit={handleSubmit}>
          <FormField name="body" rows="2" placeholder="Write a comment..." component={FormTextArea} />
          <Button
            loading={loading}
            text="Comment"
            type="submit"
            appearance="submit"
            alignWith="right"
            buttonSize="mini"
          />
      </Form>
    );
  }
}

export default compose<React.ComponentType<Props>>(
  // name of the form is passed as an attribute(form)
  // because multiple comment forms are rendered on the same page
  reduxForm({}),
)(AddCommentForm);
