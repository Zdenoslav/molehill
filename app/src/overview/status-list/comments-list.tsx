import React from 'react';
import { compose } from 'redux';
import { graphql, DataProps } from 'react-apollo';
import { Link } from 'react-router-dom';

import { StatusComments } from 'app/generated/graphql';
import { s2, s4, s5, s1 } from 'app/components/styleguide/spacing';
import styled from 'app/components/styleguide';

import Spinner from 'app/components/spinner';
import { Flex, Base } from 'app/components/styleguide/layout';
import { Body } from 'app/components/styleguide/text';

import { statusComments } from './graphql';
import UserImage from 'app/components/user-image';
import ShowMore from 'app/components/show-more';

const COMMENTS_LIMIT = 5;

const CommentsListContainer = styled(Base)`
  background-color: ${props => props.theme.backgroundDarker};
`;

const SpinnerContainer = styled(Base)`
  width: 24px;
  height: 24px;
  position: absolute;
  right: -32px;
`;

const Comment = styled(Flex)<{canLoadMore: boolean}>`
  margin-top: 8px;

  &:first-child {
    margin-top: ${props => props.canLoadMore ? 0 : '8px'};
  }
`;

const CommentsInfo = styled(Flex)`
  position: relative;
`;

const LoadMoreContainer = styled(Flex)`
  position: absolute;
  left: 50%;
  transform: translateX(-50%);
`;

const LoadMore = styled(Body)`
  color: ${props => props.theme.info};
`;

interface ICommentsList {
  statusId: string;
}

type Props = ICommentsList & DataProps<StatusComments>;

class CommentsList extends React.Component<Props> {

  constructor(props: Props) {
    super(props);
    this.handleFetchMore = this.handleFetchMore.bind(this);
  }

  public render() {
    const {
      data,
    } = this.props;

    if (!data || !data.statusComments) {
      return null;
    }

    const canLoadMore = data.statusComments.count > data.statusComments.comments.length;

    return (
      <CommentsListContainer paddingHorizontal={s5} paddingBottom={s2}>
        <CommentsInfo align="center" justify="flex-end" marginTop={s1}>
          {canLoadMore && (
            <LoadMoreContainer align="center">
              <LoadMore clickable onClick={this.handleFetchMore}>Load more</LoadMore>
              <SpinnerContainer>
                {data.loading && (
                  <Spinner />
                )}
              </SpinnerContainer>
            </LoadMoreContainer>
          )}
            <Body>{data.statusComments.count} comments</Body>
        </CommentsInfo>
        <Base>
          {[...data.statusComments.comments].reverse().map(comment => (
            <Comment key={comment.id} align="flex-start" canLoadMore={canLoadMore}>
              <Link to={`/users/${comment.user.id}`}>
                  <UserImage imgSrc={comment.user.image} />
              </Link>
              <Base marginLeft={s4}>
                <Link to={`/users/${comment.user.id}`}>
                  <Body emphasized>{comment.user.username}</Body>
                </Link>
                <ShowMore maxChars={80} textComponent={Body} text={comment.body} />
              </Base>
            </Comment>
          ))}
        </Base>
      </CommentsListContainer>
    );
  }

  private handleFetchMore() {
    const {
      data,
      statusId,
    } = this.props;

    const {
      fetchMore,
      loading,
      statusComments,
    } = data;

    if (!loading && statusComments && statusComments.comments.length) {
      fetchMore({
        variables: {
          statusId,
          limit: COMMENTS_LIMIT,
          cursor: statusComments.comments[statusComments.comments.length - 1].id,
        },
        updateQuery: (prevResult, {fetchMoreResult}) => {
          return {
            statusComments: {
              ...prevResult.statusComments,
              comments: [
                ...prevResult.statusComments.comments,
                ...fetchMoreResult.statusComments.comments,
              ],
            },
          };
        },
      });
    }
  }
}

const CommentsListWithData = compose<React.ComponentType<ICommentsList>>(
  graphql<ICommentsList, StatusComments>(statusComments, {
    options: (props) => ({
      variables: {
        statusId: props.statusId,
        limit: 3,
      },
      notifyOnNetworkStatusChange: true,
    }),
  }),
)(CommentsList);

export default CommentsListWithData;
