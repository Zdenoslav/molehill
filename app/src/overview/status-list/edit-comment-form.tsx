import React from 'react';
import { Form, InjectedFormProps, reduxForm } from 'redux-form';
import { compose } from 'redux';

import { IFormProps } from 'app/components/form-elements/typings';

import FormField from 'app/components/form-elements/form-field';
import FormTextArea from 'app/components/form-elements/form-text-area';

export interface IEditCommentFormProps {
  body: string;
}

type Props = Partial<InjectedFormProps<IEditCommentFormProps>> & IFormProps;

class EditCommentForm extends React.PureComponent<Props> {

  constructor(props: Props) {
    super(props);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  public render() {
    const {
      handleSubmit,
    } = this.props;

    return (
      <Form onSubmit={handleSubmit}>
          <FormField name="body" rows="2" component={FormTextArea} onKeyDown={this.handleKeyDown}/>
      </Form>
    );
  }

  private handleKeyDown(e: React.KeyboardEvent<any>) {
    const {
      onSubmit,
    } = this.props;

    if (onSubmit && e.keyCode === 13) {
       e.preventDefault();
       (onSubmit as any)({
        body: (e.target as any).value,
      });
    }

  }
}

export default compose<React.ComponentType<Props>>(
  reduxForm({
    form: 'EDIT_COMMENT_FORM',
    initialValues: (props: Props) => ({
      body: props.initialValues && props.initialValues.body,
    }),
  }),
)(EditCommentForm);
