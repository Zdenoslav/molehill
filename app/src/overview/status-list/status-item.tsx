import React, { RefObject } from 'react';
import { compose } from 'redux';
import { Link } from 'react-router-dom';
import { graphql, DataProps } from 'react-apollo';
import { connect } from 'react-redux';

import { Title, Body } from 'app/components/styleguide/text';
import { Flex, Base } from 'app/components/styleguide/layout';
import ShowMore from 'app/components/show-more';
import UserImage from 'app/components/user-image';
import Like from 'app/components/icons/like';

import { s5, s1, s4, s2 } from 'app/components/styleguide/spacing';
import styled, { css } from 'app/components/styleguide';
import {
  StatusesInRadius_statusesInRadius_statuses,
  AddStatusLike,
  AddStatusLikeVariables,
  UserById,
  UserByIdVariables,
  StatusesInRadius,
  StatusesInRadiusVariables,
  RemoveStatusLike,
  RemoveStatusLikeVariables,
} from 'app/generated/graphql';
import withStateMutation, { IWithStateMutationProps } from 'app/components/higher-order/with-state-mutation';
import { IRootState } from 'app/redux/root-reducer';
import { getUserId } from 'app/login/selectos';
import { userById } from 'app/user-profile/graphql';

import Comments from './comments';
import { addStatusLikeMutation, statusesInRadius, removeStatusLikeMutation } from '../graphql';
import { getRadiusInMeters } from '../selectors';
import { getLat, getLng } from '../map/selectors';

const GOOGLE_MAPS_API = 'https://www.google.com/maps/dir/?api=1&';

const buildNavigationLink = (coordinates: number[]) => {
  return `${GOOGLE_MAPS_API}destination=${coordinates[0]},${coordinates[1]}`;
};

const StatusContainer = styled(Flex)<{selected?: boolean}>`
  background: ${props => props.theme.invertedText};
  border-bottom: 1px solid ${props => props.theme.border.default};

  ${props => props.selected && css`
    border-left: 14px solid ${props => props.theme.primary};
    box-shadow: 2px 2px 4px ${props => props.theme.shadow};
    transform: translateZ(1px)
  `}
`;

const StatusContent = styled(Base)`
  border-bottom: 1px solid ${props => props.theme.border.default};
`;

const LikeIcon = styled(Like).attrs<{isLikedByUser: boolean}>({
  color: (props: any) => props.isLikedByUser ? props.theme.errorLight : props.theme.secondaryDark,
  width: 24,
  height: 24,
})``;

interface IStatusItemProps {
  isSelected: boolean;
  status: StatusesInRadius_statusesInRadius_statuses;
  selectStatus: () => void;
  forwardedRef?: RefObject<any>;
  ref?: RefObject<any> | string;
}

interface IStateProps {
  userId: string;
  radiusInMeters: number;
  userLat?: number;
  userLng?: number;
}

type Props = IStateProps & IStatusItemProps & IWithStateMutationProps<AddStatusLike, AddStatusLikeVariables>
  & IWithStateMutationProps<RemoveStatusLike, RemoveStatusLikeVariables>
  & DataProps<UserById>;

class StatusItem extends React.Component<Props, {canAddLike: boolean}> {

  constructor(props: Props) {
    super(props);
    this.handleAddStatusLike = this.handleAddStatusLike.bind(this);
    this.handleRemoveStatusLike = this.handleRemoveStatusLike.bind(this);
    this.state = {
      canAddLike: !Boolean(props.status.statusLikes.find(like => like.userId === props.userId)),
    };
  }

  public render() {
    const {
      status,
      isSelected,
      forwardedRef,
      selectStatus,
    } = this.props;

    return (
      <StatusContainer
        paddingVertical={s5}
        grow={1}
        key={status.id}
        direction="column"
        selected={isSelected}
        ref={forwardedRef}
      >
      <StatusContent paddingBottom={s2} paddingHorizontal={s5}>
        <Title
          clickable
          paddingBottom={s1}
          onClick={selectStatus}
        >
          {status.title}
        </Title>
        <a
          target="_blank"
          href={buildNavigationLink(status.location.coordinates)}
        >
          <Body disabled>{`${status.city} ${status.zipCode}, ${status.street}`}</Body>
        </a>
        <Link to={`/users/${status.user.id}`}>
          <Flex align="center" marginTop={s2}>
            <UserImage imgSrc={status.user.image} />
            <Body marginLeft={s2} marginTop={s1}>@{`${status.user.username}`}</Body>
          </Flex>
        </Link>
        {Boolean(status.description) && (
          <Base paddingTop={s4}>
            <ShowMore textComponent={Body} text={status.description} />
          </Base>
        )}
        <Flex
          isInline
          marginTop={s2}
          align="center"
          clickable
          onClick={this.state.canAddLike ? this.handleAddStatusLike : this.handleRemoveStatusLike}
        >
          <LikeIcon isLikedByUser={!this.state.canAddLike} />
          <Body marginLeft={s2}>{status.statusLikes.length}</Body>
        </Flex>
      </StatusContent>
        <Comments statusId={status.id} />
      </StatusContainer>
    );
  }

  private handleAddStatusLike() {
    const {
      status,
      addStatusLikeMutation,
      data,
      radiusInMeters,
      userLat,
      userLng,
      userId,
    } = this.props;

    this.setState({canAddLike: false}, () => {
      addStatusLikeMutation.mutate({
        variables: {
          like: {
            statusId: status.id,
            userId: data.userById!.id,
          },
        },
        // update locally cached statuses data
        update: (store, addStatusLikeResult) => {
          const statusesData = store.readQuery<StatusesInRadius, Partial<StatusesInRadiusVariables>>({
            query: statusesInRadius,
            variables: {
              radius: radiusInMeters,
              latitude: userLat as number,
              longitude: userLng as number,
              skip: false,
            },
          });

          if (statusesData && addStatusLikeResult.data) {
            const likedStatusIndex = statusesData.statusesInRadius.statuses.findIndex(({id}) => id === status.id);

            if (likedStatusIndex !== -1) {
              const likedStatus = statusesData.statusesInRadius.statuses[likedStatusIndex];
              const updatedLikedStatus = {
                ...likedStatus,
                statusLikes: [
                  ...likedStatus.statusLikes,
                  {
                    ...addStatusLikeResult.data.addStatusLike,
                    userId,
                  },
                ],
              };

              store.writeQuery<StatusesInRadius, Partial<StatusesInRadiusVariables>>({
                query: statusesInRadius,
                variables: {
                  radius: radiusInMeters,
                  latitude: userLat as number,
                  longitude: userLng as number,
                  skip: false,
                },
                data: {
                  statusesInRadius: {
                    ...statusesData.statusesInRadius,
                    statuses: [
                      ...statusesData.statusesInRadius.statuses.slice(0, likedStatusIndex),
                      updatedLikedStatus,
                      ...statusesData.statusesInRadius.statuses.slice(likedStatusIndex + 1),
                    ],
                  },
                },
              });
            }
          }
        },
      });
    });
  }

  private handleRemoveStatusLike() {
    const {
      status,
      removeStatusLikeMutation,
      radiusInMeters,
      userLat,
      userLng,
      userId,
    } = this.props;

    this.setState({canAddLike: true}, () => {
      removeStatusLikeMutation.mutate({
        variables: {
          like: {
            statusId: status.id,
            userId,
          },
        },
        // update locally cached statuses data
        update: (store, removeStatusLikeResult) => {
          const statusesData = store.readQuery<StatusesInRadius, Partial<StatusesInRadiusVariables>>({
            query: statusesInRadius,
            variables: {
              radius: radiusInMeters,
              latitude: userLat as number,
              longitude: userLng as number,
              skip: false,
            },
          });

          if (statusesData && removeStatusLikeResult.data) {
            const dislikedStatusIndex = statusesData.statusesInRadius.statuses.findIndex(({id}) => id === status.id);

            if (dislikedStatusIndex !== -1) {
              const dislikedStatus = statusesData.statusesInRadius.statuses[dislikedStatusIndex];
              const updatedDislikedStatus = {
                ...dislikedStatus,
                statusLikes: dislikedStatus.statusLikes.filter(like => like.userId !== userId),
              };

              store.writeQuery<StatusesInRadius, Partial<StatusesInRadiusVariables>>({
                query: statusesInRadius,
                variables: {
                  radius: radiusInMeters,
                  latitude: userLat as number,
                  longitude: userLng as number,
                  skip: false,
                },
                data: {
                  statusesInRadius: {
                    ...statusesData.statusesInRadius,
                    statuses: [
                      ...statusesData.statusesInRadius.statuses.slice(0, dislikedStatusIndex),
                      updatedDislikedStatus,
                      ...statusesData.statusesInRadius.statuses.slice(dislikedStatusIndex + 1),
                    ],
                  },
                },
              });
            }
          }
        },
      });

    });
  }

}

export default compose<React.ComponentType<IStatusItemProps>>(
  connect<IStateProps, {}, {}, IRootState>((state) => ({
    userId: getUserId(state),
    radiusInMeters: getRadiusInMeters(state),
    userLat: getLat(state),
    userLng: getLng(state),
  })),
  graphql<IStateProps & IStatusItemProps, UserById, UserByIdVariables>(userById, {
    options: (props) => ({
      variables: {
        id: props.userId,
      },
    }),
  }),
  graphql(addStatusLikeMutation, {name: 'addStatusLikeMutation'}),
  withStateMutation({name: 'addStatusLikeMutation'}),
  graphql(removeStatusLikeMutation, {name: 'removeStatusLikeMutation'}),
  withStateMutation({name: 'removeStatusLikeMutation'}),
)(React.forwardRef((props: any, ref: RefObject<any>) => {
  return <StatusItem {...props} forwardedRef={ref} />;
}));
