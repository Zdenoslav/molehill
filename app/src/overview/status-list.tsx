import React, { createRef } from 'react';
import { findDOMNode } from 'react-dom';
import { compose } from 'redux';
import { graphql } from 'react-apollo';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { Title, Body } from 'app/components/styleguide/text';
import { Flex, Base } from 'app/components/styleguide/layout';
import Spinner from 'app/components/spinner';
import ShowMore from 'app/components/show-more';
import UserImage from 'app/components/user-image';

import { s5, s1, s4, s2 } from 'app/components/styleguide/spacing';
import styled, { css } from 'app/components/styleguide';
import { IRootState } from 'app/redux/root-reducer';
import { getStopAutoRefetchStatuses } from './map/selectors';
import { NAVBAR_HEIGHT } from 'app/components/navbar';

import { statusesInRadius } from './graphql';
import { StatusesInRadiusData } from './types';
import { getSelectedStatusId, getRadiusInMeters } from './selectors';
import { selectStatus } from './actions';

const GOOGLE_MAPS_API = 'https://www.google.com/maps/dir/?api=1&';

const buildNavigationLink = (coordinates: number[]) => {
  return `${GOOGLE_MAPS_API}destination=${coordinates[0]},${coordinates[1]}`;
};

const StatusItem = styled(Flex)<{selected?: boolean}>`
  background: ${props => props.theme.invertedText};
  border-bottom: 1px solid ${props => props.theme.border.default};

  ${props => props.selected && css`
    border-left: 14px solid ${props => props.theme.primary};
    box-shadow: 2px 2px 4px ${props => props.theme.shadow};
    transform: translateZ(1px)
  `}
`;

const StatusListContainer = styled.div`
  box-shadow: 1px 0 3px ${props => props.theme.border.focus};
`;

interface IStatusListProps {
  userLat?: number;
  userLng?: number;
  className?: string;
}

interface IStateProps {
  selectedStatusId: string;
  radius: number;
  stopAutoRefetchStatuses: boolean;
}

type Props = StatusesInRadiusData & IStatusListProps & IStateProps;

class StatusList extends React.Component<Props> {

  private selectedItemRef: React.RefObject<any>;

  // Used for checking if status was selected by clicking on status title
  // so the selected item is not scrolled if it's already in the view
  private nextSelectedStatusId: string;

  constructor(props: Props) {
    super(props);
    this.selectedItemRef = createRef();
  }

  public componentDidUpdate(prevProps: Props) {
    if (prevProps.selectedStatusId !== this.props.selectedStatusId &&
      this.nextSelectedStatusId !== this.props.selectedStatusId
    ) {
      if (this.selectedItemRef.current) {
        const statusItemNode = findDOMNode(this.selectedItemRef.current);
        if (statusItemNode) {
          const {offsetTop} = statusItemNode as HTMLElement;
          window.scrollTo(0, offsetTop - NAVBAR_HEIGHT);
        }
      }
    }
  }

  public shouldComponentUpdate(nextProps: Props) {
    // don't update if radius has been changed
    if (this.props.stopAutoRefetchStatuses && nextProps.stopAutoRefetchStatuses) {
      return false;
    }
    return true;
  }

  public render() {
    const {
      data,
      className,
      selectedStatusId,
      stopAutoRefetchStatuses,
    } = this.props;

    return (
      <StatusListContainer className={className}>
        {!data || data.loading || stopAutoRefetchStatuses ? (
          <Flex padding={s5} direction="column" align="center" justify="center">
            <Spinner />
            <Body marginTop={s2}>Loading...</Body>
          </Flex>
        ) : (
          data.statusesInRadius && data.statusesInRadius.length ? (
            data.statusesInRadius.map(status => (
              <StatusItem
                padding={s5}
                grow={1}
                key={status.id}
                direction="column"
                selected={status.id === selectedStatusId}
                ref={status.id === selectedStatusId ? this.selectedItemRef : ''}
              >
                <Title
                  clickable
                  paddingBottom={s1}
                  onClick={() => this.handleSelectedStatusChanged(status.id)}
                >
                  {status.title}
                </Title>
                <a
                  target="_blank"
                  href={buildNavigationLink(status.location.coordinates)}
                >
                  <Body disabled>{`${status.city} ${status.zipCode}, ${status.street}`}</Body>
                </a>
                <Link to={`/users/${status.user.id}`}>
                  <Flex align="center" marginTop={s2}>
                    <UserImage imgSrc={status.user.image} />
                    <Body marginLeft={s2} marginTop={s1}>@{`${status.user.username}`}</Body>
                  </Flex>
                </Link>
                <Base paddingTop={s4}>
                  <ShowMore textComponent={Body} text={status.description} />
                </Base>
              </StatusItem>
            ))
          ) : (
            <Body padding={s5} disabled>No statuses found. Be first to share the moment with people near you.</Body>
          )
        )}
      </StatusListContainer>
    );
  }

  private handleSelectedStatusChanged(statusId: string) {
    this.nextSelectedStatusId = statusId;
    selectStatus.dispatch(statusId);
  }
}

export default compose<React.ComponentType<IStatusListProps>>(
  connect<IStateProps, {}, StatusesInRadiusData & IStatusListProps, IRootState>(
    (state) => ({
      selectedStatusId: getSelectedStatusId(state),
      radius: getRadiusInMeters(state),
      stopAutoRefetchStatuses: getStopAutoRefetchStatuses(state),
    }),
  ),
  graphql<IStatusListProps & IStateProps, StatusesInRadiusData>(statusesInRadius, {
    options: (props) => ({
      variables: {
        radius: props.radius,
        latitude: props.userLat,
        longitude: props.userLng,
        skip: props.stopAutoRefetchStatuses,
      },
    }),
    skip: ({userLat, userLng}) => !userLat || !userLng,
  }),
)(StatusList);
