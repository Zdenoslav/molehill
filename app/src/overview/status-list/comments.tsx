import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { graphql, DataProps } from 'react-apollo';
import { reset } from 'redux-form';

import withStateMutation, { IWithStateMutationProps } from 'app/components/higher-order/with-state-mutation';
import { getUserId, getUsername } from 'app/login/selectos';
import { IRootState } from 'app/redux/root-reducer';
import { updateError } from 'app/components/global-event/actions';
import store from 'app/redux/store';
import { s5 } from 'app/components/styleguide/spacing';

import { Base } from 'app/components/styleguide/layout';

import AddCommentForm, { IAddCommentFormProps } from './add-comment-form';
import CommentsList from './comments-list';
import { addCommentMutation, statusComments } from './graphql';
import { StatusComments, StatusCommentsVariables, UserById, AddComment } from 'app/generated/graphql';
import { userById } from 'app/user-profile/graphql';

interface ICommentsProps {
  statusId: string;
}

interface IStateProps {
  userId: string;
  username: string;
}

type Props = IStateProps & IWithStateMutationProps & ICommentsProps & DataProps<UserById>;

class Comments extends React.PureComponent<Props> {

  private formName: string;

  constructor(props: Props) {
    super(props);
    this.handleAddComment = this.handleAddComment.bind(this);
    this.formName = `status-${props.statusId}`;
  }

  public render() {
    const {
      sMutation,
      statusId,
    } = this.props;

    return (
      <>
        <CommentsList statusId={statusId} />
        <Base paddingHorizontal={s5}>
          <AddCommentForm loading={sMutation.loading} form={this.formName} onSubmit={this.handleAddComment}/>
        </Base>
      </>
    );
  }

  private handleAddComment(values: IAddCommentFormProps) {
    const {
      sMutation,
      userId,
      statusId,
      data,
    } = this.props;

    const newComment = {
      body: values.body,
      userId,
      statusId,
    };

    sMutation.mutate({
      variables: {
        comment: newComment,
      },
      // update locally cached comments data
      update: (store, addCommentResult) => {
        const commentsData = store.readQuery<StatusComments, StatusCommentsVariables>({
          query: statusComments,
          variables: {
            statusId,
          },
        });

        if (commentsData && data.userById) {
          store.writeQuery<StatusComments, StatusCommentsVariables>({
            query: statusComments,
            variables: {
              statusId,
            },
            data: {
              statusComments: {
                comments: [
                  {
                    ...addCommentResult.data.addComment,
                    ...newComment,
                    user: data.userById,
                  },
                  ...commentsData.statusComments.comments,
                ],
               count: commentsData.statusComments.count + 1,
               __typename: 'StatusCommentsWithCount',
              },
            },
          });
        }
      },
    })
    .then(() => {
      store.dispatch(reset(this.formName));
    })
    .catch(e => updateError.dispatch('Failed to add a comment'));
  }
}

export default compose(
  connect<IStateProps, {}, ICommentsProps, IRootState>((state) => ({
    userId: getUserId(state),
    username: getUsername(state),
  })),
  graphql<IStateProps & ICommentsProps>(userById, {
    options: (props) => ({
      variables: {
        id: props.userId,
      },
    }),
  }),
  graphql<IStateProps & ICommentsProps, AddComment>(addCommentMutation),
  withStateMutation(),
)(Comments);
