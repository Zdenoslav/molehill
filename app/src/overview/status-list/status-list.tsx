import React, { createRef } from 'react';
import { findDOMNode } from 'react-dom';
import { compose } from 'redux';
import { graphql } from 'react-apollo';
import { connect } from 'react-redux';

import { Body } from 'app/components/styleguide/text';
import Spinner from 'app/components/spinner';

import { s2, s4 } from 'app/components/styleguide/spacing';
import { IRootState } from 'app/redux/root-reducer';
import { NAVBAR_HEIGHT } from 'app/components/navbar';
import styled from 'app/components/styleguide';

import { getStopAutoRefetchStatuses } from '../map/selectors';
import { statusesInRadius } from '../graphql';
import { StatusesInRadiusData } from '../types';
import { getSelectedStatusId, getRadiusInMeters, getSelectedCategoryIds } from '../selectors';
import { selectStatus } from '../actions';

import StatusItem from './status-item';
import { StatusesInRadiusVariables } from 'app/generated/graphql';
import Container from 'app/components/container';
import LikesModal from './status-likes/likes-modal';
import { getLikeStatusId } from '../status-modal/selectors';

const STATUSES_LIMIT = 10;

const StyledContainer = styled(Container)`
  background-color: ${props => props.theme.backgroundDarker};
`;

interface IStatusListProps {
  userLat?: number;
  userLng?: number;
  className?: string;
}

interface IStateProps {
  selectedStatusId: string;
  radius: number;
  selectedCategories?: string[];
  stopAutoRefetchStatuses: boolean;
  likeStatusId: string;
}

type Props = StatusesInRadiusData & IStatusListProps & IStateProps;

class StatusList extends React.Component<Props> {

  private selectedItemRef: React.RefObject<any>;

  // Used for checking if status was selected by clicking on status title
  // so the selected item is not scrolled if it's already in the view
  private nextSelectedStatusId: string;

  constructor(props: Props) {
    super(props);
    this.selectedItemRef = createRef();
    this.handleFetchMore = this.handleFetchMore.bind(this);
  }

  public componentDidUpdate(prevProps: Props) {
    if (prevProps.selectedStatusId !== this.props.selectedStatusId &&
      this.nextSelectedStatusId !== this.props.selectedStatusId
    ) {
      if (this.selectedItemRef.current) {
        const statusItemNode = findDOMNode(this.selectedItemRef.current);
        if (statusItemNode) {
          const {offsetTop} = statusItemNode as HTMLElement;
          window.scrollTo(0, offsetTop - NAVBAR_HEIGHT);
        }
      }
    }
  }

  public shouldComponentUpdate(nextProps: Props) {
    // don't update if radius has been changed
    if (this.props.stopAutoRefetchStatuses && nextProps.stopAutoRefetchStatuses) {
      return false;
    }
    return true;
  }

  public render() {
    const {
      data,
      className,
      selectedStatusId,
      stopAutoRefetchStatuses,
      likeStatusId,
    } = this.props;

    const canLoadMore = data.statusesInRadius && data.statusesInRadius.count > data.statusesInRadius.statuses.length;

    return (
      <StyledContainer withShadow noPadding className={className} direction="column">
        {!data || data.loading || stopAutoRefetchStatuses ? (
          <Container direction="column" align="center" justify="center">
            <Spinner />
            <Body marginTop={s2}>Loading...</Body>
          </Container>
        ) : (
          data.statusesInRadius && data.statusesInRadius.statuses.length ? (
            data.statusesInRadius.statuses.map(status => (
              <StatusItem
                key={status.id}
                isSelected={status.id === selectedStatusId}
                status={status}
                ref={status.id === selectedStatusId ? this.selectedItemRef : ''}
                selectStatus={() => this.handleSelectedStatusChanged(status.id)}
              />
            ))
          ) : (
            <Container>
              <Body disabled>No statuses found. Share the moment with people near you.</Body>
            </Container>
          )
        )}
        {canLoadMore && (
          <Container clickable padding={s4} justify="center" onClick={this.handleFetchMore}>
            <Body>Load more</Body>
          </Container>
        )}
        <LikesModal statusId={likeStatusId} header="People who like it" />
      </StyledContainer>
    );
  }

  private handleSelectedStatusChanged(statusId: string) {
    this.nextSelectedStatusId = statusId;
    selectStatus.dispatch(statusId);
  }

  private handleFetchMore() {
    const {
      data,
      radius,
      userLat,
      userLng,
    } = this.props;

    const {
      fetchMore,
      loading,
      statusesInRadius,
    } = data;

    if (!loading && statusesInRadius && statusesInRadius.statuses.length) {
      fetchMore({
        variables: {
          radius,
          latitude: userLat as number,
          longitude: userLng as number,
          skip: false,
          limit: STATUSES_LIMIT,
          cursor: statusesInRadius.statuses[statusesInRadius.statuses.length - 1].id,
        },
        updateQuery: (prevResult, {fetchMoreResult}) => {
          return {
            statusesInRadius: {
              ...prevResult.statusesInRadius,
              statuses: [
                ...prevResult.statusesInRadius.statuses,
                ...fetchMoreResult.statusesInRadius.statuses,
              ],
            },
          };
        },
      });
    }
  }
}

export default compose<React.ComponentType<IStatusListProps>>(
  connect<IStateProps, {}, StatusesInRadiusData & IStatusListProps, IRootState>(
    (state) => ({
      selectedStatusId: getSelectedStatusId(state),
      radius: getRadiusInMeters(state),
      stopAutoRefetchStatuses: getStopAutoRefetchStatuses(state),
      likeStatusId: getLikeStatusId(state),
      selectedCategories: getSelectedCategoryIds(state),
    }),
  ),
  graphql<IStatusListProps & IStateProps, StatusesInRadiusData, StatusesInRadiusVariables>(statusesInRadius, {
    options: (props) => ({
      variables: {
        radius: props.radius,
        categoryIds: props.selectedCategories,
        latitude: props.userLat as number,
        longitude: props.userLng as number,
        skip: props.stopAutoRefetchStatuses,
        limit: 5,
      },
    }),
    skip: ({userLat, userLng}) => !userLat || !userLng,
  }),
)(StatusList);
