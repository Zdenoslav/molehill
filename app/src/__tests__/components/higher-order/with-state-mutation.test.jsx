import React from 'react';
import { shallow } from 'enzyme';

import {StatusItem} from 'app/overview/status-list/status-item';
import withStateMutation from 'app/components/higher-order/with-state-mutation';

describe('withStateMutation', () => {

  it('should render withStateMutation', () => {
    const WithStateMutationElement = withStateMutation()(StatusItem);
    const wrapper = shallow(<WithStateMutationElement mutate={jest.fn()} />);
    expect(wrapper).toMatchSnapshot();
  });

});
