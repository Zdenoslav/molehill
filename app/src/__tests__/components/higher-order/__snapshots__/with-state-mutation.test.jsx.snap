// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withStateMutation should render withStateMutation 1`] = `
<Fragment
  mutate={[Function]}
  sMutation={
    Object {
      "error": null,
      "loading": false,
      "mutate": [Function],
      "result": null,
    }
  }
/>
`;
